File: 01_Arrays_&_Hashing/2_Medium/LongestConsecutiveSubsequence.py
'''
Given an unsorted array of integers nums, 
return the length of the longest consecutive 
elements sequence.
'''

def longestConsecutive(nums):
    unique = set(nums)
    longest = 0 

    for num in nums:
        if num - 1 not in unique:
            l = 0
            while n + l in unique:
                l += 1
            
            longest = max(longest, l)
        
    
    return longest 

# The start of a subsequence will be such a number such that 
# num - 1 is not in the array. We store the unique numbers in a 
# set for constant time lookup. Once we get to such a number by 
# traversing the array, we loop until l such that num + l is not
# in the array. This is the length of this consecutive subsequence. 
# We keep track of the maximum length. 

# Time | Space: O(n) | O(n)

File: 01_Arrays_&_Hashing/02_Medium/LongestConsecutiveSubsequence.cpp
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums){
        unordered_set<int> unique;
        int longest = 0;

        for (int num : nums){
            if (unique.find(num - 1) == unique.end()){
                int l = 0;
                while (unique.find(num + 1) != unique.end()){
                    l++;
                };
                longest = max(longest,l);
            };
        };

        return longest;
    };
};